adoc/chapters/cuda_backend.adoc

```diff
@@ -0,0 +1,598 @@
+not yet supported.
+
+==== Work Item Mapping
+
+A SYCL `nd_range` will tranpose indices as it maps to hardware memory.
+This gives better memory access patterns, in general.
+
+SYCL uses row major memory ordering, meaning in some memory object the 
+rows will be contiguous in memory. SYCL follows C++ convention in 
+this regard. Following the row-major paradigm, it is intuitive to imagine 
+each work-item in a `parallel_for` indexing through a contiguous block of 
+memory as it does its work. However, this gives poorly coalesced memory 
+accesses, as a given contiguous chunk of data being loaded may only pass 
+memory to a single work-item. More efficient memory access patterns are 
+achieved when each load of contiguous data can give data to as many 
+work-items as possible. Meaning the data used by a given work-item
+is non-contiguous.
+
+SYCL makes this intuitive row-major C++ approach give good memory access 
+patterns by flipping the indices of the `nd_range`, as it maps to hardware.
+
+The linear id (whose use is not recommended) of a two dimensional `nd_range` 
+can be calculated using:
+
+[source,c++]
+----
+cgh.parallel_for(range<2>(64, 128), [=](item<2> it) {
+  size_t linearIdx = it.get_id(1) + (it.get_id(0) * it.get_range(0));
+  ...
+});
+----
+
+Notice that rows appear to be accessed in a column-major, rather than 
+row-major, format. This is only the case because the indices are flipped 
+by the SYCL implementation. All memory in SYCL is stored in row-major format.
+
+It is best to avoid calculating the linear index manually; it is better
+to use a multi-dimensional `sycl::id` to index into memory, as it doesn't
+expose index-flipping to the user.
+
+[[table.cuda.CUDA_features_to_SYCL]]
+.CUDA execution features with their corresponding SYCL features
+[width="100%",options="header",cols="50%,50%"]
+|====
+| [code]#SYCL#                                                       | [code]#CUDA#
+| [code]#Compute unit#                                               | [code]#Streaming multiprocessor#
+| [code]#nd_range#                                                   | [code]#grid#
+| [code]#work-group#                                                 | [code]#Thread block#
+| [code]#sub-group#                                                  | [code]#thread_block_tile#
+| [code]#work-item#                                                  | [code]#Thread#
+| [code]#SYCL nd_item synchronization#                               | [code]#cudaStreamSynchronize#
+| [code]#work-group barrier#                                         | [code]#__syncthread#
+|====
+
+[[sec::programming_interface]]
+== Programming Interface
+
+[[sub:cuda:queries]]
+=== Queries
+
+For all event information profiling descriptors, the calls to 
+[code]#sycl::event::get_profiling_info# return the time (in nanoseconds) since 
+the creation of the context that the event is associated with. The "Resolution"
```

Comment:
@t4c1 what do you think?

BTW should the meaning of "timebase" be obvious? The interpretation for it that I get from #242 is that it represents "timestamp" from an event that has the same invariant value across a given backend. I thought this should also be satisfied in the CUDA backend with the caveat of the large 10^-6 error range. BTW do other backends have such large errors with respect to the nanosecond unit time? Should a note that the error can be much larger than 10^-9 in some backends be added to the content from #242?

Reply:
The DPC++ implementation can be changed to use a common base event, for example, in the platform object. So it can conform to the changes in the SYCL spec, but the CUDA backend spec needs to be updated.

Reply:
I don't think the spec needs to say anything about the granularity of the timestamps.  This seems like a "quality of implementation" thing.

Reply:
I think it is OK for this CUDA backend specification to note the timer resolution, so  it is OK to keep this wording:

> The resolution (timing error) of the returned value is the same as that provided by the CUDA driver API call `cuEventElapsedTime`: +/- 0.5 microseconds.

Does everyone agree that this statement (and the implementation) will change, though?

> For all event information profiling descriptors, the calls to  `sycl::event::get_profiling_info` return the time (in nanoseconds) **since the creation of the context** that the event is associated with.

Instead, we want all profiling timestamps from the CUDA backend to share the same "timebase".  It sounds like you think this is possible to implement, @t4c1?

Reply:
Yes, it should not be too much work.

Reply:
@t4c1 can you specify what the exact wording of the update should be now?

Reply:
Something like: 
```suggestion
the creation of the CUDA platform (which happens when application is started). The "Resolution"
```